PMM-5738 Basic core

percona-csalguero opened the pull request.

percona-csalguero committed:
PMM-5738 Basic core

BupycHuk commented on main.go:
```diff
@@ -1,13 +1,63 @@
 import (
+	"fmt"
+
 	"github.com/Percona-Lab/mnogo_exporter/exporter"
```
please move it below as separate block

AlekSi replied:
Issues like that should break CI. Both goimports configuration and `git diff --exit-code` are missing

AlekSi commented on .goreleaser.yml:
```diff
@@ -0,0 +1,34 @@
+# Make sure to check the documentation at http://goreleaser.com
+before:
+  hooks:
+    # you may remove this if you don't use vgo
```
There is no `vgo` and should not be any "you may" or "this is an example" – this should not be an example.

AlekSi commented on .goreleaser.yml:
```diff
@@ -0,0 +1,34 @@
+    # you may remove this if you don't need go generate
+    - go generate ./...
+builds:
+  - 
```
I think we should reformat the whole file.

AlekSi commented on Makefile:
```diff
@@ -0,0 +1,97 @@
+	@grep '^[a-zA-Z]' $(MAKEFILE_LIST) | \
+	awk -F ':.*?## ' 'NF==2 {printf "  %-26s%s\n", $$1, $$2}'
+
+test: env  						## Run all tests
```
Comment line indentations are not aligned.

AlekSi commented on docker/Dockerfile.common:
```diff
@@ -0,0 +1,26 @@
+
+USER nobody
+
+CMD ["pbmctl"]
```
The whole file is not about MongoDB exporter. Is it even used?

AlekSi commented on docker/test/gen-certs/mongodb-test-ca.crt:
```diff
@@ -0,0 +1,39 @@
+-----BEGIN CERTIFICATE-----
```
Do we really have to keep all those files in git? Can we generate them on depend with gen-certs.sh or (preferably) with mkcert?

percona-csalguero replied:
certs generated with mkcert are not working. It seems it is because mkcert is not setting the DNS for ip addresses. Added auto generation with the current script

AlekSi commented on exporter/exporter.go:
```diff
@@ -2,25 +2,58 @@ package exporter
+// Opts holds new exporter options.
 type Opts struct {
 	DSN string
+	Log *logrus.Logger
```
Doesn't seem to be used?

AlekSi commented on exporter/exporter.go:
```diff
@@ -2,25 +2,58 @@ package exporter
 )
 
+const (
+	connectTimeout = 10 * time.Second
```
Isn't that a part of DSN?

AlekSi commented on exporter/exporter_test.go:
```diff
@@ -2,25 +2,80 @@ package exporter
 	require.NoError(t, err)
 
-	t.Cleanup(func() { client.Disconnect(ctx) })
+	t.Cleanup(func() { client.Disconnect(ctx) }) //nolint:errcheck
```
Given that we already have a closure, let's check returned error

percona-csalguero committed:
PMM-5738 Fixes for CR

percona-csalguero committed:
PMM-5738 Fixes for github actions test

percona-csalguero committed:
PMM-5738 Updated github actions tests

percona-csalguero committed:
PMM-5738 cosmetic change in makefile

percona-csalguero committed:
PMM-5738 cosmetic change in github action file

percona-csalguero committed:
PMM-5738 Testing build

percona-csalguero committed:
PMM-5738 Updated mod

percona-csalguero committed:
PMM-5738 Github actions test

percona-csalguero committed:
PMM-5738 Updated mod

percona-csalguero committed:
PMM-5738 Updated mod

percona-csalguero committed:
PMM-5738 Fixed tests

percona-csalguero committed:
PMM-5738 Updated makefile

artemgavrilov commented on .goreleaser.yml:
```diff
@@ -0,0 +1,29 @@
+        - -s -w -X main.version={{.Env.VERSION}} -X main.commit={{.ShortCommit}} -X main.buildDate={{.Date}}
+archives:
+  -
+    name_template: "{{ .ProjectName }}-{{ .Env.VERSION }}.{{ .Os }}-{{ .Arch }}"
```
Seems like that should be single line

artemgavrilov commented on Makefile:
```diff
@@ -0,0 +1,106 @@
+GO_TEST_CODECOV?=
+
+TOP_DIR=$(shell git rev-parse --show-toplevel)
+VERSION ?=$(shell git describe --abbrev=0)
```
Mixed style
```
VAR?=$value
VAR ?=$value
```

I suggest third variant ```VAR ?= $value```, IMHO it's easier to read

artemgavrilov commented on Makefile:
```diff
@@ -0,0 +1,106 @@
+
+FILES = $(shell find . -type f -name '*.go' -not -path "./vendor/*")
+
+init:						## Install linters
```
Indent such comments with spaces instead of tabs, please. Different editors visualize tabs differently.

Also, some comments have missing dot at the end.

percona-csalguero committed:
PMM-5738 Fixes for CR

- Aligned help
- Formated var definitions
- Ran linter for make and added missing .PHONY entries

percona-csalguero committed:
PMM-5738 Updated dependencies

percona-csalguero committed:
PMM-5738 Replaced tabs by spaces in comments

percona-csalguero committed:
PMM-5738 Updated deps

BupycHuk commented on .goreleaser.yml:
```diff
@@ -0,0 +1,28 @@
+    ldflags: 
+        - -s -w -X main.version={{.Env.VERSION}} -X main.commit={{.ShortCommit}} -X main.buildDate={{.Date}}
+archives:
+  - name_template: "{{ .ProjectName }}-{{ .Env.VERSION }}.{{ .Os }}-{{ .Arch }}"
```
this and name_template below looks different.

BupycHuk commented on docker-compose.yml:
```diff
@@ -1,8 +1,135 @@
-    ports:
-      - 127.0.0.1:27017:27017
+  standalone:
+    network_mode: host
```
why do all instances here have `network_mode: host`?

AlekSi replied:
⚠️ Yeah, that will not work Mac, and we expect other developers to work on this exporter on Macs.

BupycHuk commented on docker-compose.yml:
```diff
@@ -1,8 +1,135 @@
+    - ./docker/test/mongod.key:/mongod.key:ro
+    - ./docker/test/ssl/rootCA.crt:/rootCA.crt:ro
+    - ./docker/test/ssl/mongodb.pem:/mongod.pem:ro
+#  configsvr2:
```
Do we need them commented?

percona-csalguero replied:
A real cluster has 3 config servers. Just 1 works so to not to have that many server instances for nothing, I commented out 2 of 3

AlekSi commented on .github/workflows/go.yml:
```diff
@@ -22,15 +22,23 @@ jobs:
-            curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
-            dep ensure
-        fi
+        go get golang.org/x/tools/cmd/goimports
```
That is already done by `make init`. So we only have to move "Get dependencies" (not only linters) up.

AlekSi commented on Makefile:
```diff
@@ -0,0 +1,105 @@
+
+
+init:                       ## Install linters.
+	curl https://raw.githubusercontent.com/reviewdog/reviewdog/master/install.sh| sh -s
```
Please add space before `|`

percona-csalguero committed:
PMM-5738 New sandbox

percona-csalguero committed:
PMM-5738 Fixed make

percona-csalguero committed:
Updated sum

BupycHuk commented on docker-compose.yml:
```diff
@@ -1,8 +1,107 @@
 ---
-version: "3.7"
+version: '2'
```
Why version 2? and why we need so much docker-compose files?

percona-csalguero replied:
I consolidated all dockerfiles in one

BupycHuk replied:
there are still a lot of docker-compose files, please remove other ones if they are not required or add comment why they are needed.

percona-csalguero replied:
done

percona-csalguero committed:
PMM-5738 Consolitated all dockerfiles

percona-csalguero committed:
PMM-5738 Consolitated all dockerfiles

percona-csalguero committed:
Updated deps

BupycHuk commented on docker/scripts/.gitattributes:
```diff
@@ -0,0 +1 @@
+* text=auto eol=lf
```
do we need this?

percona-csalguero replied:
removed

percona-csalguero committed:
PMM-5738 Deleted ignored fileds

percona-csalguero committed:
PMM-5738 Updated deps

AlekSi commented on .github/workflows/go.yml:
```diff
@@ -22,15 +21,22 @@ jobs:
-            dep ensure
-        fi
+        make format
+        git diff --exit-code
```
I think we could move that to the very end to check that tests do not modify repo state too

AlekSi commented on docker/scripts/setup-cnf.sh:
```diff
@@ -0,0 +1,40 @@
+    };
+    rs.initiate(cfg, { force: true });
+    rs.reconfig(cfg, { force: true });
+EOF
```
I think GitHub is doing a good job there – let's add \n

AlekSi commented on docker/scripts/setup.sh:
```diff
@@ -0,0 +1,39 @@
+    };
+    rs.initiate(cfg, { force: true });
+    rs.reconfig(cfg, { force: true });
+EOF
```
\n

AlekSi commented on exporter/exporter.go:
```diff
@@ -7,20 +7,43 @@ import (
+}
+
+// Disconnect from the database.
+func (e *Exporter) Disconnect(ctx context.Context) error {
```
When we will call this method?

If we are going to support both "exporter is always connected" and "exporter re-connects on each scrape" modes, then client can't be a field of the shared object.

percona-csalguero replied:
removed

AlekSi commented on main.go:
```diff
@@ -1,13 +1,63 @@
-		panic(err)
+		log.Fatal(err)
 	}
+
```
❓ Ok, and?.. The intention was to build an actual exporter – without any MongoDB metrics / collectors, but with default metrics provided by the Prometheus library (go process metrics, a number of scrapes, etc.) on `/metrics` path. That would also show how the connection will be established, and whatever it will be shared or not.

AlekSi replied:
It looks like part of it is done at #3. I think we should move it there – not the actual collector, but code that starts exporter and gathers data from the future collectors.

percona-csalguero committed:
PMM-5738 WIP

percona-csalguero committed:
PMM-5738 Fixes for CR

percona-csalguero committed:
PMM-5738 Fixes for CR

percona-csalguero committed:
PMM-5738 Updated deps

artemgavrilov commented on Makefile:
```diff
@@ -0,0 +1,96 @@
+	@echo $(TEST_ENV) | tr ' ' '\n' >.env
+
+init:                       ## Install linters.
+	curl https://raw.githubusercontent.com/reviewdog/reviewdog/master/install.sh | sh -s
+	curl https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s latest
+	go get golang.org/x/tools/cmd/goimports
+	# To prevent goimports being marked as dependency
+	go mod tidy
```
On SaaS we download all such tools as sources and use `tools.go` to pin versions
```
init:                   ## Install development tools
	go build -modfile=tools/go.mod -o bin/goimports golang.org/x/tools/cmd/goimports
	go build -modfile=tools/go.mod -o bin/golangci-lint github.com/golangci/golangci-lint/cmd/golangci-lint
	go build -modfile=tools/go.mod -o bin/reviewdog github.com/reviewdog/reviewdog/cmd/reviewdog
```

It guarantees that each developer has same tools versions. If you want I can make a PR to show how we use that approach.

percona-csalguero replied:
Yes please.

artemgavrilov replied:
Here it is: https://github.com/Percona-Lab/mnogo_exporter/pull/5

artemgavrilov commented on Makefile:
```diff
@@ -0,0 +1,96 @@
+build:                      ## Build the binaries.
+	goreleaser --snapshot --skip-publish --rm-dist
+
+FILES = $(shell find . -type f -name '*.go' -not -path "./vendor/*")
+format:                     ## Format source code.
+	go mod tidy
+	gofmt -w -s $(FILES)
+	goimports -local github.com/Percona-Lab/mnogo_exporter -l -w $(FILES)
```
That project hasn't vendor, so it looks like you can use 
 ```
	gofmt -w -s .
	goimports -local github.com/Percona-Lab/mnogo_exporter -l -w .
```

percona-csalguero committed:
PMM-5738 Updated tools

percona-csalguero committed:
PMM-5738 Updated tools

percona-csalguero committed:
PMM-5738 Updated tools

AlekSi commented on exporter/exporter.go:
```diff
@@ -2,25 +2,83 @@ package exporter
+
+	gatherers := prometheus.Gatherers{}
+	gatherers = append(gatherers, prometheus.DefaultGatherer)
+	gatherers = append(gatherers, registry)
```
```go
gatherers := prometheus.Gatherers{prometheus.DefaultGatherer, registry}
```
